Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1393,6 +1403,28 @@ public enum ResolvedUnaryOp { BoolNot, BitwiseNot, Cardinality }
     /// reference types or subset types thereof.
     /// </summary>
     protected abstract void EmitTypeTest(string localName, Type fromType, Type toType, IToken tok, ConcreteSyntaxTree wr);
+
+    /// <summary>
+    /// Emit a conjunct that tests if the Dafny real number "source" is an integer, like:
+    ///    "TestIsInteger(source) && "
+    /// It is fine for the target code to repeat the mention of "source", if necessary.
+    /// </summary>
+    protected virtual void EmitIsIntegerTest(Expression source, ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) { }
+
+    /// <summary>
+    /// Emit a conjunct that tests if the Dafny integer "source" is a character, like:
+    ///     "TestIsRune(source) && "
+    /// It is fine for the target code to repeat the mention of "source", if necessary.
+    /// </summary>
+    protected virtual void EmitIsRuneTest(Expression source, ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) { }
```

Comment:
Why isn't this `EmitIsCharacterTest`? I assume it's semantics change with the value of `--unicode-char`?

Reply:
I see from the PR description this is only for `--unicode-char:true` mode, which is fine. But I'd still avoid "rune", since that term isn't part of the Unicode standard, and in fact can be confused with the Runic code block (https://en.wikipedia.org/wiki/Runic_(Unicode_block)). Or in this case can be misread as something specific to C# or Rust which DO use that term.

I'd go with "Unicode scalar value" instead.

Reply:
Thanks. I made the renaming in a separate PR: https://github.com/dafny-lang/dafny/pull/5115.